jetson/nav/recovery.cpp

```diff
@@ -0,0 +1,120 @@
+#include "recovery.hpp"
+#include "utilities.hpp"
+
+Recovery::Recovery(int offset, double dispThresh) :
+                    offsetAmount(offset), dispThresh(dispThresh){
+
+}//Recovery()
+
+bool Recovery::isStuck(){
+    // check displacement (TODO: have condition for if we are turning! or just don't call if we are turning?)
+    // if displacement < threshold: return true
+    // else return false
+
+    double distance = estimateDistance(odoms.front().odom,odoms.back().odom);
+    if (!odoms.empty() && getDuration(odoms.front().time,odoms.back().time) >= offsetAmount && distance < dispThresh){
+        return true;
+    }
+    else {
+        return false;
+    }
+
+
+}//isStuck()
+
+void Recovery::update(Odometry currOdom, time_point currTime){
+    // updates odoms deque with new data
+    // and remove older data
+    // TODO: add fail safe for overloading deque with data and making program fail (so if )
+
+    // add latest odom
+    OdomTimePoint newOdom = {currOdom,currTime};
+    odoms.push_front(newOdom);
+
+    // check old odoms for removal (if newer odoms that are within offset time)
+    // check oldest odom (and second to last odom to see if we should remove the oldest odom)
+    while ( odoms.size() > 2 && (getDuration(odoms.front().time,odoms[odoms.size()-2].time)) > offsetAmount){
+        odoms.pop_back();
+    }
+
+}//update()
+
+
+int Recovery::getDuration(time_point p1, time_point p2){
+    // get duration between two points in seconds (rounded to nearest second)
+    // let p1 be the newer of the time points and p2 be the later of the time points
+    // for our use case
+    auto milliseconds = std::chrono::duration_cast<std::chrono::milliseconds>(p1 - p2);
+    return int((milliseconds.count()/1000.0) + 0.5);
```

Comment:
you can duration_cast to seconds instead of milliseconds and use `count()`. Also why is this /2 ?

Reply:
Unsure what you mean by /2 but this should be resolved now